packages/components/src/image-cropper/reducer.ts

```diff
@@ -0,0 +1,286 @@
+/**
+ * Internal dependencies
+ */
+import type { Position, Size } from './types';
+import { rotatePoint, degreeToRadian, getFurthestVector } from './math';
+
+export type State = {
+	// The container/image's width.
+	width: number;
+	// The container/image's height.
+	height: number;
+	// The rotation angle between -45deg to 45deg.
+	angle: number;
+	// The number of 90-degree turns.
+	turns: 0 | 1 | 2 | 3;
+	// The zoom scale.
+	scale: number;
+	// Whether the image is flipped horizontally.
+	flipped: boolean;
+	// The offset position of the cropper window.
+	offset: Position;
+	// The position of center of the image.
+	position: Position;
+	// The size of the cropper window.
+	size: Size;
+	// Whether the cropper window is resizing.
+	isResizing: boolean;
+	// Whether the image is dragging/moving.
+	isDragging: boolean;
+};
```

Comment:
A good aim could be to model the state like this:
```js
State = {
    image: {
		// the image would always have a `top`/`left` of zero, because we apply `transform.translation` to move it, but a rect is a friendly storage format.
        rect: { top, right, bottom, left, x, y, width, height }
    },
    imageTransforms: {
        rotation: number,
        scale: number, 
        translation: Vector2<number>,
        reflect: Vector2<bool>
    },
    // the resizable box
    clippingRect: { 
        rect: { top, right, bottom, left, x, y, width, height }
    },
    // UI states
    isResizing: bool,
    isDragging: bool,
}
```

I'm visualizing the image manipulation as two separate things. 

Firstly, if you completely disregard the 'clipping rect', there are distinct changes the user can make to the image. They can rotate it (slider), zoom it (mouse wheel) and translate it (drag it). As the user makes those changes the deltas are added to the transform state. The `image` state doesn't change at all until the user clicks 'apply', at which point the `transform` state is reset to zero.

Keeping the transform as a state like this is pretty important as the scaling and rotation need to be applied before translation when transforming something geometrically around its center (assuming the origin is the center of the image).

The clipping rect is a separate thing on top of that, but from what I can tell, it doesn't affect the underlying image until the user applies the crop, so I think a `rect` is the only state needed for it.

There might be more needed after that, for example I notice that the clipping rect and the image are often translated together, but it'd be better to store that as a separate `transform` object (e.g. `viewTransform`).

This reflects pretty closely how video games work in having a series of 'transforms', though they store them in matrices:
https://gamedev.stackexchange.com/questions/178643/the-view-matrix-finally-explained

Reply:
I was also going to comment that `turns` might be more portable/easily understood as a `rotate` value, possibly in combination with the `deg/radian`?

Reply:
Do you mean a `rotate` and a `degree` value? I never like the naming of `turns` so I might switched to that 😅.

Reply:
> Do you mean a rotate and a degree value? 

Oh, the controls are separate, right? Then, yes 😄